### PR TITLE
Added roles implementation

### PR DESCRIPTION
**Pull request by MilianCode**

### CODE REVIEW

File: src/main/java/com/miliancode/model/Role.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/Role.java]:
```suggestion
@Getter
@Setter
```

File: src/main/java/com/miliancode/model/Role.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+@Table(name = "roles")
+public class Role implements GrantedAuthority {
+    private static final String PREFIX = "ROLE_";
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/Role.java]: remove prefix

File: src/main/java/com/miliancode/model/Role.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(unique = true, nullable = false, name = "name")
+    private String roleNameString;
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/Role.java]:
```suggestion
    @Column(nullable = false, unique = true)
    @Enumerated(EnumType.STRING)
    private RoleName name;
```

File: src/main/java/com/miliancode/model/Role.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Override
+    public String getAuthority() {
+        return PREFIX + roleNameString;
+    }
+
+    public RoleName getRoleName() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/Role.java]: remove get and set

File: src/main/java/com/miliancode/model/Role.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(unique = true, nullable = false, name = "name")
+    private String roleNameString;
+
+    @Override
+    public String getAuthority() {
+        return PREFIX + roleNameString;
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/Role.java]:
```suggestion
        return name.name();

```

File: src/main/java/com/miliancode/model/User.java [hunk 6]
```diff
@@ -30,6 +40,44 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
-    @Column(nullable = false)
+    @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
+    @ToString.Exclude
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/com/miliancode/model/User.java [hunk 7]
```diff
@@ -30,6 +40,44 @@ public class User {
...
+    @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
+    @ToString.Exclude
+    @ManyToMany(fetch = FetchType.EAGER)
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/com/miliancode/model/User.java [hunk 8]
```diff
@@ -30,6 +40,44 @@ public class User {
...
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return new HashSet<>(roles);
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/User.java]:
```suggestion
        return roles;
```

File: src/main/java/com/miliancode/repository/RoleRepository.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.miliancode.repository;
+
+import com.miliancode.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Role findByRoleNameString(String roleNameString);
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/repository/RoleRepository.java]:
```suggestion
    Role findByName(Role.RoleName name);
```

File: src/main/java/com/miliancode/security/SecurityConfig.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
+package com.miliancode.security;
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/security/SecurityConfig.java]:
```suggestion
package com.miliancode.config;
```

File: src/main/java/com/miliancode/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -4,24 +4,35 @@
...
+    private final RoleRepository roleRepository;
 
     @Override
     public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
             throw new RegistrationException("User with this email is already registered");
         }
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Set<Role> roles = new HashSet<>();
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/service/impl/UserServiceImpl.java]:
```suggestion
```

File: src/main/java/com/miliancode/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -4,24 +4,35 @@
...
+        roles.add(roleRepository.findByRoleNameString(Role.RoleName.USER.toString()));
+        user.setRoles(roles);
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/service/impl/UserServiceImpl.java]:
```suggestion
        user.setRoles(Set.of(userRole));
```

File: src/main/java/com/miliancode/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -4,24 +4,35 @@
...
 
     @Override
     public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
             throw new RegistrationException("User with this email is already registered");
         }
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Set<Role> roles = new HashSet<>();
+        roles.add(roleRepository.findByRoleNameString(Role.RoleName.USER.toString()));
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/service/impl/UserServiceImpl.java]:
```suggestion
        Role userRole = roleRepository.findByName(Role.RoleName.ROLE_USER);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve it

File: src/main/java/com/miliancode/model/User.java (hunk 6 above)
Comment (MilianCode) [reply src/main/java/com/miliancode/model/User.java]: Without it, I receiving LazyInitializationException when trying to do any operation with books: failed to lazily initialize a collection of role: com.miliancode.model.User.roles: could not initialize proxy - no Session

File: src/main/java/com/miliancode/model/Role.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "roles")
+public class Role implements GrantedAuthority {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/model/Role.java]:
```suggestion
```

File: src/main/java/com/miliancode/model/User.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/main/java/com/miliancode/model/User.java]: Eeager is a bad approach, we have some options to resolve such exception, like 
`@EntityGraph(attributePaths = "roles")` under the method in UserRepo 
or
you can use join fetch in query (not preferable)

File: src/main/java/com/miliancode/service/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -4,24 +4,33 @@
...
     @Override
     public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
             throw new RegistrationException("User with this email is already registered");
         }
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Role userRole = roleRepository.findByName(Role.RoleName.ROLE_USER);
+        user.setRoles(Set.of(userRole));
         User savedUser = userRepository.save(user);
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
```

File: src/main/java/com/miliancode/service/impl/UserServiceImpl.java [hunk 16]
```diff
@@ -4,24 +4,33 @@
...
         return userMapper.toUserResponse(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/com/miliancode/service/impl/UserServiceImpl.java]:
```suggestion
        return userMapper.toUserResponse(user);

```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**